File: packages/shared/schemas/ticket.ts
import { z } from "zod";

export const CreateTicketSchema = z.object({
  userId: z.string(),
  subject: z.string(),
  description: z.string(),
  status: z.string(),
  priority: z.string(),
  labels: z.string(),
  assigneeId: z.string(),
  lastReplyAt: z.string().nullable(),
  closedAt: z.string().nullable(),
  user: z.unknown(),
  assignee: z.unknown(),
  replies: z.unknown(),
});

export const UpdateTicketSchema = z.object({
  userId: z.string().optional(),
  subject: z.string().optional(),
  description: z.string().optional(),
  status: z.string().optional(),
  priority: z.string().optional(),
  labels: z.string().optional(),
  assigneeId: z.string().optional(),
  lastReplyAt: z.string().nullable(),
  closedAt: z.string().nullable(),
  user: z.unknown().optional(),
  assignee: z.unknown().optional(),
  replies: z.unknown().optional(),
});

export type CreateTicketInput = z.infer<typeof CreateTicketSchema>;
export type UpdateTicketInput = z.infer<typeof UpdateTicketSchema>;

File: apps/api/internal/models/ticket.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Ticket — single support request. Anyone authenticated can open one;
// ADMIN/EDITOR roles see the full queue, regular USER role sees their
// own. Status is open by default; closing stamps ClosedAt.
type Ticket struct {
	ID          string         `gorm:"primarykey;size:36" json:"id"`
	UserID      string         `gorm:"size:36;index;not null" json:"user_id"`
	Subject     string         `gorm:"size:200;not null" json:"subject"`
	Description string         `gorm:"type:text;not null" json:"description"`
	Status      string         `gorm:"size:16;index;default:'open'" json:"status"`   // open | closed
	Priority    string         `gorm:"size:16;index;default:'medium'" json:"priority"` // low | medium | high | critical
	Labels      string         `gorm:"size:255" json:"labels"`                           // comma-separated up to 8
	AssigneeID  string         `gorm:"size:36;index" json:"assignee_id"`                // optional, must be ADMIN role
	LastReplyAt *time.Time     `json:"last_reply_at"`                                    // touched by every reply
	ClosedAt    *time.Time     `json:"closed_at"`
	CreatedAt   time.Time      `gorm:"index" json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`

	// Eager-loaded relations for the detail page.
	User     *User           `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Assignee *User           `gorm:"foreignKey:AssigneeID" json:"assignee,omitempty"`
	Replies  []TicketReply   `gorm:"foreignKey:TicketID" json:"replies,omitempty"`
}

func (t *Ticket) BeforeCreate(tx *gorm.DB) error {
	if t.ID == "" {
		t.ID = uuid.New().String()
	}
	if t.Status == "" {
		t.Status = "open"
	}
	if t.Priority == "" {
		t.Priority = "medium"
	}
	return nil
}

// TicketReply — chronological thread under a ticket. IsAdminReply lets
// the dashboard style admin replies differently (badged, opposite side
// of the thread, etc.).
type TicketReply struct {
	ID            string         `gorm:"primarykey;size:36" json:"id"`
	TicketID      string         `gorm:"size:36;index;not null" json:"ticket_id"`
	UserID        string         `gorm:"size:36;index;not null" json:"user_id"`
	Body          string         `gorm:"type:text;not null" json:"body"`
	IsAdminReply  bool           `json:"is_admin_reply"`
	CreatedAt     time.Time      `gorm:"index" json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	DeletedAt     gorm.DeletedAt `gorm:"index" json:"-"`

	User *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

func (r *TicketReply) BeforeCreate(tx *gorm.DB) error {
	if r.ID == "" {
		r.ID = uuid.New().String()
	}
	return nil
}
